### PR TITLE
[REVIEW] Add `cudf.set_allocator` function for easier RMM config

### PR DESCRIPTION
**Pull request by shwina**
Without memory pool:

```python
In [1]: import cudf                                                                                                                                                                           

In [2]: import numpy as np                                                                                                                                                                    

In [3]: a = cudf.Series(np.arange(100_000_000))                                                                                                                                               

In [4]: pos = cudf.Series(np.arange(50_000_000))                                                                                                                                              

In [5]: %timeit a[pos]                                                                                                                                                                        
32.6 ms ± 18.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

With memory pool (default allocator), and initial pool size of 1/2 GPU memory:

```python
In [1]: import cudf                                                                                                                                                                           

In [2]: import numpy as np                                                                                                                                                                    

In [3]: cudf.set_allocator("default", pool=True)                                                                                                                                              

In [4]: a = cudf.Series(np.arange(100_000_000))                                                                                                                                               

In [5]: pos = cudf.Series(np.arange(50_000_000))                                                                                                                                              

In [6]: %timeit a[pos]                                                                                                                                                                        
9.14 ms ± 324 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

With memory pool (default allocator), and initial pool size of 0 bytes (i.e., *caching* allocator):

```python
In [1]: import cudf                                                                                                                                                                           

In [2]: import numpy as np                                                                                                                                                                    

In [3]: cudf.set_allocator(pool=True, initial_pool_size=0)                                                                                                                                    

In [4]: a = cudf.Series(np.arange(100_000_000))                                                                                                                                               

In [5]: pos = cudf.Series(np.arange(50_000_000))                                                                                                                                              

In [6]: %timeit a[pos] 
10.5 ms ± 221 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

cc: @jrhemstad @harrism @jakirkham

### CODE REVIEW

File: python/cudf/cudf/__init__.py [hunk 1]
```diff
@@ -30,6 +30,62 @@
...
+            initial_pool_size = 1  # Since "0" is semantic value, use 1 byte
+        if not isinstance(initial_pool_size, int):
+            raise TypeError("initial_pool_size must be an integer")
+        _rmm_cfg.initial_pool_size = initial_pool_size
+    _rmm_cfg.enable_logging = enable_logging
+    _rmm.initialize()
+
+
+@_initfunc
+def set_allocator(allocator="default", pool=False, initial_pool_size=None):
```
Comment (harrism) [inline python/cudf/cudf/__init__.py]: Please add the logging option that RMM provides.

Comment (harrism) [reply python/cudf/cudf/__init__.py]: Ah, didn't see that. That works.

File: python/cudf/cudf/__init__.py [hunk 2]
```diff
@@ -30,6 +30,65 @@
...
+        ``"default"`` : use default allocator.
+        ``"managed"`: use managed memory allocator.
+    pool : bool
+        Enable memory pool.
+    initial_pool_size : int
+        Memory pool size in bytes. If ``None`` (default), 1/2 of total
+        GPU memory is used. If ``pool=False``, this argument is ignored.
+    """
+    use_managed_memory = True if allocator == "managed" else False
+    _set_rmm_config(use_managed_memory, pool, initial_pool_size)
```
Comment (kkraus14) [inline python/cudf/cudf/__init__.py]: I think we should move these functions to a different file, maybe utils or somewhere else and then import them here instead.

File: python/cudf/cudf/__init__.py [hunk 3]
```diff
@@ -30,6 +30,65 @@
...
+    use_managed_memory : bool, optional
+        If ``True``, use cudaMallocManaged as underlying allocator.
+        If ``False`` (default), use  cudaMalloc.
+    use_pool_allocator : bool
+        If ``True``, enable pool mode.
+        If ``False`` (default), disable pool mode.
+    initial_pool_size : int, optional
+        If ``use_pool_allocator=True``, sets initial pool size.
+        If ``None``, us
+es 1/2 of total GPU memory.
```
Comment (kkraus14) [inline python/cudf/cudf/__init__.py]: Erroneous line break?

**Comment by kkraus14** [comment]
`python/cudf/cudf/utils/utils.py:278:1: W293 blank line contains whitespace`

File: python/cudf/cudf/utils/utils.py [hunk 4]
```diff
@@ -212,3 +230,60 @@ def __array_function__(self, *args, **kwargs):
...
+    rmm_cfg.enable_logging = enable_logging
+    rmm.initialize()
+
+
+@initfunc
+def set_allocator(allocator="default", pool=False, initial_pool_size=None):
+    """
+    Set the GPU memory allocator. This function should be run only once,
+    before any cudf objects are created.
+
```
Comment (harrism) [inline python/cudf/cudf/utils/utils.py]:
```suggestion

```

Comment (harrism) [reply python/cudf/cudf/utils/utils.py]: @shwina if you leave the little "allow administrators to commit to this branch" box checked when you create your PRs, we can fix little details like this for you while you sleep, reducing PR latency during the endgame...

Comment (jakirkham) [reply python/cudf/cudf/utils/utils.py]: I think we are able to push commits to the branch based on this line below.

```
Add more commits by pushing to the add-set-allocator branch on shwina/cudf.
```

That said, I don't think GitHub allows us to do this through the suggestion UI if we are not the PR author unfortunately. Have run into this before. 😞

FWIW also tried to add the suggested change, but it didn't let me do this either. 🤷‍♂

Comment (kkraus14) [reply python/cudf/cudf/utils/utils.py]: Yea, github doesn't allow you to commit suggestions for some reason. You can still edit the file via the github UI or push commits directly to the branch without issue though.

**Comment by shwina** [comment]
Thanks @kkraus14 :)

**Comment by jakirkham** [comment]
Thanks all! Will this then be included in the nightlies?

**Comment by harrism** [comment]
Yep.

**Pull request merged**